pageserver/src/tenant.rs

```diff
@@ -1244,32 +1249,50 @@ impl Tenant {
-        let mut timelines = self.timelines.lock().unwrap();
+    pub async fn delete_timeline(&self, timeline_id: TimelineId) -> anyhow::Result<()> {
+        // Transition the timeline into TimelineState::Paused
+        let timeline = {
+            let mut timelines = self.timelines.lock().unwrap();
 
-        // Ensure that there are no child timelines **attached to that pageserver**,
-        // because detach removes files, which will break child branches
-        let children_exist = timelines
-            .iter()
-            .any(|(_, entry)| entry.get_ancestor_timeline_id() == Some(timeline_id));
+            // Ensure that there are no child timelines **attached to that pageserver**,
+            // because detach removes files, which will break child branches
+            let children_exist = timelines
+                .iter()
+                .any(|(_, entry)| entry.get_ancestor_timeline_id() == Some(timeline_id));
 
-        anyhow::ensure!(
-            !children_exist,
-            "Cannot delete timeline which has child timelines"
-        );
-        let timeline_entry = match timelines.entry(timeline_id) {
-            Entry::Occupied(e) => e,
-            Entry::Vacant(_) => bail!("timeline not found"),
+            anyhow::ensure!(
+                !children_exist,
+                "Cannot delete timeline which has child timelines"
+            );
+            let timeline_entry = match timelines.entry(timeline_id) {
+                Entry::Occupied(e) => e,
+                Entry::Vacant(_) => bail!("timeline not found"),
+            };
+
+            let timeline = Arc::clone(timeline_entry.get());
+            timeline.set_state(TimelineState::Paused);
+
+            drop(timelines);
+            timeline
         };
 
-        let timeline = timeline_entry.get();
-        timeline.set_state(TimelineState::Paused);
+        // XXX this call can panic if GC or Compaction are still running.
+        // It seems like the assumption is that they aren't because this function
+        // panics if it can't try_lock() the underlying mutex.
+        // That assumption is generally wrong because GC / Compaction that was
+        // started before transitioning the timeline into `TimelineState::Paused` might
+        // still be running at this point.
+        //
+        // => Find a proper solution to wait until this timeline is only referenced by us,
+        //    and no other thread will act on it every again.
+        let layer_removal_guard = timeline.layer_removal_guard().await?;
 
-        let layer_removal_guard = timeline.layer_removal_guard()?;
+        // NB: storage_sync upload tasks that reference these layers have been cancelled
+        //     by the caller.
 
         let local_timeline_directory = self.conf.timeline_path(&timeline_id, &self.tenant_id);
+        // XXX make this atomic so that, if we crash-mid-way, the timeline won't be picked up
+        // with some layers missing.
```

Comment:
Could e.g rename the directory it to have `.temp` suffix first. (Doesn't need to be part of this PR as it's a pre-existing problem)

Reply:
Yeah, I though about renaming it to the uninit suffix, so that on next pageserver restart, we clean up.